sklearn/utils/stats.py

```diff
@@ -15,47 +18,143 @@ def _weighted_percentile(array, sample_weight, percentile=50):
     sorted_weights = _take_along_axis(sample_weight, sorted_idx, axis=0)
+    percentile = np.array([percentile / 100] * n_cols)
+    cum_weigths = stable_cumsum(sorted_weights, axis=0)
+
+    def _squeeze_arr(arr, n_dim):
+        return arr[0] if n_dim == 1 else arr
+
+    # Percentile can be computed with 3 different alternative:
+    # https://en.wikipedia.org/wiki/Percentile
+    # These 3 alternatives depend of the value of a parameter C. NumPy uses
+    # the variant where C=0 which allows to obtained a strictly monotically
+    # increasing function which is defined as:
+    # P = (x - 1) / (N - 1); x in [1, N]
+    # Weighted percentile change this formula by taking into account the
+    # weights instead of the data frequency.
+    # P_w = (x - w) / (S_w - w), x in [1, N], w being the weight and S_w being
+    # the sum of the weights.
+    adjusted_percentile = (cum_weigths - sorted_weights)
+    with np.errstate(invalid="ignore"):
+        adjusted_percentile /= cum_weigths[-1] - sorted_weights
+        nan_mask = np.isnan(adjusted_percentile)
+        adjusted_percentile[nan_mask] = 1
+
+    if interpolation in ("lower", "higher", "nearest"):
+        percentile_idx = np.array([
+            np.searchsorted(adjusted_percentile[:, col], percentile[col],
+                            side="left")
+            for col in range(n_cols)
+        ])
+
+        if interpolation == "lower" and np.all(percentile < 1):
+            # P = 100 is a corner case for "lower"
+            percentile_idx -= 1
+        elif interpolation == "nearest" and np.all(percentile < 1):
+            for col in range(n_cols):
+                error_higher = abs(
+                    adjusted_percentile[percentile_idx[col], col] -
+                    percentile[col]
+                )
+                error_lower = abs(
+                    adjusted_percentile[percentile_idx[col] - 1, col] -
+                    percentile[col]
+                )
+                if error_higher >= error_lower:
+                    percentile_idx[col] -= 1
+
+        percentile_idx = np.apply_along_axis(
+            lambda x: np.clip(x, 0, n_rows - 1), axis=0,
+            arr=percentile_idx
+        )
+
+        percentile_value = array[
+            sorted_idx[percentile_idx, np.arange(n_cols)],
+            np.arange(n_cols)
+        ]
+        percentile_value = _squeeze_arr(percentile_value, n_dim)
+
+    else:  # interpolation == "linear"
+        percentile_value = np.array([
+            np.interp(
+                x=percentile[col],
+                xp=adjusted_percentile[:, col],
+                fp=array[sorted_idx[:, col], col],
+            )
+            for col in range(n_cols)
+        ])
+
+        percentile_value = _squeeze_arr(percentile_value, n_dim)
+
+    single_sample_weight = np.count_nonzero(sample_weight, axis=0)
+    if np.any(single_sample_weight == 1):
+        # edge case where a single weight is non-null in which case the
```

Comment:
do we have a strong incentive to support this?

Reply:
It was in the original test and not supporting it would be a regression indeed.